cmd/earthly/main.go

```diff
@@ -412,10 +417,27 @@ func newEarthlyApp(ctx context.Context, console conslogging.ConsoleLogger) *eart
 		},
 		&cli.StringFlag{
 			Name:        "buildkit-host",
+			Value:       buildkitd.DockerAddress,
 			EnvVars:     []string{"EARTHLY_BUILDKIT_HOST"},
 			Usage:       wrap("The URL to use for connecting to a buildkit host. ", "If empty, earthly will attempt to start a buildkitd instance via docker run"),
 			Destination: &app.buildkitHost,
 		},
+		&cli.IntFlag{
+			Name:        "buildkit-port",
+			Value:       config.DefaultBuildkitPort,
+			EnvVars:     []string{"EARTHLY_BUILDKIT_PORT"},
+			Usage:       wrap("The port to use for connecting to a buildkit host. ", "Used in combination with buildkit-host to produce the URL to dial."),
+			Destination: &app.buildkitPort,
+			Hidden:      true,
+		},
```

Comment:
I feel like these are hard to follow because the host and port exist in two separate fields. It's a bit unclear what happens in some edge cases like if the user wanted to connect to a host without a port number.

I would do the configuration a little differently:

- `buildkit-host` remains like it is today: a full URL.
- `buildkit-port` turns into something like `buildkit-listening-port` and is used for configuring the buildkit daemon that we manage. We can choose a default port for this. Say `8372`, like you chose here.

The different possible cases:
* When `buildkit-host` is of the form `docker-container://` and it matches the container name we expect, again, we manage the daemon ourselves, but we use the old connectivity. (This should be the default)
* When `buildkit-host`'s port matches `tcp://127.0.0.1:<buildkit-listening-port>`, we manage the daemon ourselves and use TCP.
* When `buildkit-host` is anything else, we don't manage the daemon ourselves - we assume that the user did what was necessary to bring this it up separately.

So to turn on the feature, you would simply do

```
earthly --buildkit-host=127.0.0.1:8372
```

I think that this is a little simpler because the full address is not comprised of multiple flags and you can repoint earthly to a different buildkit with a single flag override.

Reply:
The tl;dr:
* Config is structured this way because I wanted to align how we also deal with the address we connect to for the debugger. I wanted to make it consistent, and also impossible for the user to accidentally let these diverge. Letting the hosts vary implies that you can debug from a different host than what buildkitd is running on, and I thought we weren't into that.
* I would rather keep a simple boolean for enabling TCP than requiring an address. The ports, and even the host root, are entirely optional (as they would be if we make TCP the default, which I hope we do). The options are there if you want to customize (which would likely only be for remote buildkits, or local port clash. It makes the flagged feature behave more like it would if it were productionized).

I think theres a middle ground here. Perhaps this is palatable:
* Drop the k8s-like port + host config style in favor of two addresses. Warn (but don't stop?) if the hosts diverge.
Keep the --use-tcp option, and have it simply swap the default address... more or less like it does now, and would likely in the future. Its essentially short for --buildkit-host tcp://127.0.0.1:8372 , which is the 80% case.
* Old behavior reigns if --use-tcp is not specified. Pinned town more completely:
  * Old buildkit host setting is honored, we just simply try to connect to it.
  * A new debug host setting would be ignored completely in favor of the old port setting. Port setting is depricated when the feature flag is on.
  * --use-tcp turns into a config-file only setting.